crypto/hash/htr/hashtree.go

```diff
@@ -1,17 +1,47 @@
 package htr
 
 import (
+	"runtime"
+	"sync"
+
 	"github.com/prysmaticlabs/gohashtree"
 )
 
+const minSliceSizeToParallelize = 5000
+
+func hashParallel(inputList [][32]byte, outputList [][32]byte, wg *sync.WaitGroup) {
+	defer wg.Done()
+	err := gohashtree.Hash(outputList, inputList)
+	if err != nil {
+		panic(err)
+	}
+}
+
 // VectorizedSha256 takes a list of roots and hashes them using CPU
 // specific vector instructions. Depending on host machine's specific
 // hardware configuration, using this routine can lead to a significant
 // performance improvement compared to the default method of hashing
 // lists.
-func VectorizedSha256(inputList [][32]byte, outputList [][32]byte) {
-	err := gohashtree.Hash(outputList, inputList)
+func VectorizedSha256(inputList [][32]byte) [][32]byte {
+	outputList := make([][32]byte, len(inputList)/2)
+	if len(inputList) < minSliceSizeToParallelize {
+		err := gohashtree.Hash(outputList, inputList)
+		if err != nil {
+			panic(err)
+		}
+		return outputList
+	}
+	wg := sync.WaitGroup{}
+	n := runtime.GOMAXPROCS(0)
+	wg.Add(n - 1)
+	groupSize := len(inputList) / (2 * n)
```

Comment:
why is the group size (2*n) , if this is arbitrary parameter, maybe we can set 2 as a  constant which we can change if we want to later.

Reply:
It's not arbitrary, this is the number of chunks that each processor will hash